cli/lint.rs

```diff
@@ -161,21 +177,106 @@ fn lint_file(file_path: PathBuf) -> Result<Vec<LintDiagnostic>, ErrBox> {
   Ok(file_diagnostics)
 }
 
-fn format_diagnostic(d: &LintDiagnostic) -> String {
-  let pretty_message =
-    format!("({}) {}", colors::gray(&d.code), d.message.clone());
-
-  fmt_errors::format_stack(
-    true,
-    &pretty_message,
-    Some(&d.line_src),
-    Some(d.location.col as i64),
-    Some((d.location.col + d.snippet_length) as i64),
-    &[fmt_errors::format_location(
-      &d.location.filename,
-      d.location.line as i64,
-      d.location.col as i64,
-    )],
-    0,
-  )
+trait LintReporter {
+  fn visit(&mut self, d: &LintDiagnostic);
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox);
+  fn close(&mut self);
+}
+
+#[derive(Serialize)]
+struct LintError {
+  file_path: String,
+  message: String,
+}
+
+struct PrettyLintReporter {
+  lint_count: u32,
+}
+
+impl PrettyLintReporter {
+  fn new() -> PrettyLintReporter {
+    PrettyLintReporter { lint_count: 0 }
+  }
+}
+
+impl LintReporter for PrettyLintReporter {
+  fn visit(&mut self, d: &LintDiagnostic) {
+    self.lint_count += 1;
+
+    let pretty_message =
+      format!("({}) {}", colors::gray(&d.code), d.message.clone());
+
+    let message = fmt_errors::format_stack(
+      true,
+      &pretty_message,
+      Some(&d.line_src),
+      Some(d.location.col as i64),
+      Some((d.location.col + d.snippet_length) as i64),
+      &[fmt_errors::format_location(
+        &d.location.filename,
+        d.location.line as i64,
+        d.location.col as i64,
+      )],
+      0,
+    );
+
+    eprintln!("{}\n", message);
+  }
+
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox) {
+    eprintln!("Error linting: {}", file_path);
+    eprintln!("   {}", err);
+  }
+
+  fn close(&mut self) {
+    match self.lint_count {
+      1 => eprintln!("Found 1 problem"),
+      n if n > 1 => eprintln!("Found {} problems", self.lint_count),
+      _ => (),
+    }
+  }
+}
+
+#[derive(Serialize)]
+struct JsonLintReporter {
+  diagnostics: Vec<LintDiagnostic>,
+  errors: Vec<LintError>,
+}
+
+impl JsonLintReporter {
+  fn new() -> JsonLintReporter {
+    JsonLintReporter {
+      diagnostics: Vec::new(),
+      errors: Vec::new(),
+    }
+  }
+}
+
+impl LintReporter for JsonLintReporter {
+  fn visit(&mut self, d: &LintDiagnostic) {
+    self.diagnostics.push(d.clone());
+  }
+
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox) {
+    self.errors.push(LintError {
+      file_path: file_path.to_string(),
+      message: err.to_string(),
+    });
+  }
+
+  fn close(&mut self) {
+    // Sort so that we guarantee a deterministic output which is useful for tests
+    self
+      .diagnostics
+      .sort_by(|a, b| get_sort_key(&a).cmp(&get_sort_key(&b)));
+
+    let json = serde_json::to_string_pretty(&self);
+    println!("{}", json.unwrap());
```

Comment:
**Heads up!** I just updated the `JsonLintReporter` to print to stdout instead of stderr so that we can easily pipe the output `deno lint --unstable --json | pbcopy`. Seemed pretty helpful and I couldn't find a reason why printing to `stderr` would be better in the json case 🤔

Reply:
> I couldn't find a reason why printing to `stderr` would be better in the json case 🤔

Because that's what it does without `--json` and it is much simpler to treat `--json` as strictly a different output format. If you _were_ to change it to stdout, you'd have to change the exit code to be unconditionally 0 since it clearly no longer treats diagnostics as "errors".

Reply:
> If you were to change it to stdout, you'd have to change the exit code to be unconditionally 0 since it clearly no longer treats diagnostics as "errors".

It looks like we've already been doing this https://github.com/denoland/deno/pull/6940/files#diff-8dcab4b639b85d952afd2523e77a0b3dR80. 🤔 

I'm fine either way, but writing to stdout is a nice convenience and seems to be what other linting CLI's do (i.e. `eslint`). We can handle this in a follow-up issue if you'd prefer :shrug:

Reply:
> It looks like we've already been doing this https://github.com/denoland/deno/pull/6940/files#diff-8dcab4b639b85d952afd2523e77a0b3dR80. 🤔

No? The line you linked says we haven't. Since you're not just talking about `--json`, we've already discussed this: #6303. `eslint` is not the only existing linter and probably not the one we want to copy.

Please revert this change.

Reply:
> Please revert this change.

No problem!

> No? The line you linked says we haven't.

Ah! I see. I misunderstood what you were saying in the first place. I haven't interpreted stderr and error codes this way before but I don't have a strong opinion either way. Reverted the commit 👍